lib/eth/key/decrypter.rb

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2016-2022 The Ruby-Eth Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "json"
+require "scrypt"
+
+class Eth::Key::Decrypter
+  include Eth::Util
+
+  # Class method `Eth::Key::Decrypter.perform`
+  #
+  # @return [String] decrypted key
+  def self.perform(data, password)
+    new(data, password).perform
+  end
+
+  # Constructor of the `Eth::Key::Decrypter` class for secret key
+  # encryption.
+  #
+  # @param data [JSON] encryption data including cypherkey
+  # @param password [String] password to decrypt the key
+  def initialize(data, password)
+    @data = JSON.parse(data)
+    @password = password
+  end
+
+  # Method to decrypt key using password
+  #
+  # @return [String] decrypted key
+  def perform
+    derive_key password
+    check_macs
+    bin_to_hex decrypted_data
+  end
+
+  private
+
+  attr_reader :data, :key, :password
+
+  def derive_key(password)
+    case kdf
+    when "pbkdf2"
+      @key = OpenSSL::PKCS5.pbkdf2_hmac(password, salt, iterations, key_length, digest)
+    when "scrypt"
+      # OpenSSL 1.1 inclues OpenSSL::KDF.scrypt, but it is not available usually, otherwise we could do: OpenSSL::KDF.scrypt(password, salt: salt, N: n, r: r, p: p, length: key_length)
+      @key = SCrypt::Engine.scrypt(password, salt, n, r, p, key_length)
```

Comment:
This is fine. if you want to keep the comment about OpenSSL for future reference, please consider a line break :)